[c10d] Template computeLengthsAndOffsets()

jiayisuse opened the pull request:
Summary: Different backends accept different type of length to, like MPI_Alltoallv, nccSend/Recv(), gloo::alltoallv(). So to make computeLengthsAndOffsets() template

Test Plan:
Sandcastle
CI
HPC: ./trainer_cmd.sh -p 16 -n 8 -d nccl

Differential Revision: D22961459

osalpekar commented on torch/lib/c10d/Utils.hpp:
```diff
@@ -356,6 +356,74 @@ std::vector<T*> getDataPointers(const std::vector<at::Tensor>& tensors) {
   return ptrs;
 }
 
+inline void checkSplitSizes(
```
Should we add some brief comments explaining what these helper functions do?

osalpekar reviewed (approved):
LGTM! Stamped Internally as well

jiayisuse committed:
[c10d] Template computeLengthsAndOffsets() (#42706)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42706

Different backends accept different type of length to, like MPI_Alltoallv, nccSend/Recv(), gloo::alltoallv(). So to make computeLengthsAndOffsets() template

Test Plan:
Sandcastle
CI
HPC: ./trainer_cmd.sh -p 16 -n 8 -d nccl

Reviewed By: osalpekar

Differential Revision: D22961459

fbshipit-source-id: 6ad43e9fc123213961841d46b49e1fcb9b89d618
